crates/typst-syntax/src/lexer.rs

```diff
@@ -29,6 +29,10 @@ pub(super) enum LexMode {
     Markup,
     /// Math atoms, operators, etc.
     Math,
+    /// An argument in a math function.
+    MathArg,
+    /// Function argument list in math.
+    MathArgs,
```

Comment:
Following your comment in the PR description, I do think it would be ideal to use #4364's work on returning a SyntaxNode instead of creating new modes if possible. But we'll have to wait and see how that PR progresses.